File: RUBiS-1.4.3/EJB_EntityBean_id/edu/rice/rubis/beans/servlets/PutComment.java
package edu.rice.rubis.beans.servlets;

import java.io.IOException;

import javax.naming.Context;
import javax.naming.InitialContext;
import javax.rmi.PortableRemoteObject;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import edu.rice.rubis.beans.Item;
import edu.rice.rubis.beans.ItemHome;
import edu.rice.rubis.beans.ItemPK;
import edu.rice.rubis.beans.User;
import edu.rice.rubis.beans.UserHome;
import edu.rice.rubis.beans.UserPK;

/** This servlets display the page allowing a user to put a comment
 * on an item.
 * It must be called this way :
 * <pre>
 * http://..../PutComment?to=ww&itemId=xx&nickname=yy&password=zz
 *    where ww is the id of the user that will receive the comment
 *          xx is the item id
 *          yy is the nick name of the user
 *          zz is the user password
 * </pre>
 * @author <a href="mailto:cecchet@rice.edu">Emmanuel Cecchet</a> and <a href="mailto:julie.marguerite@inrialpes.fr">Julie Marguerite</a>
 * @version 1.0
 */


public class PutComment extends HttpServlet
{
 

  private void printError(String errorMsg, ServletPrinter sp)
  {
    sp.printHTMLheader("RUBiS ERROR: PutComment");
    sp.printHTML("<h2>Your request has not been processed due to the following error :</h2><br>");
    sp.printHTML(errorMsg);
    sp.printHTMLfooter();
  }


  /**
   * Describe <code>doGet</code> method here.
   *
   * @param request a <code>HttpServletRequest</code> value
   * @param response a <code>HttpServletResponse</code> value
   * @exception IOException if an error occurs
   * @exception ServletException if an error occurs
   */
  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException
  {
    ServletPrinter sp = null;
    String toStr = request.getParameter("to");
    String itemStr = request.getParameter("itemId");
    String name = request.getParameter("nickname");
    String pass = request.getParameter("password");
    sp = new ServletPrinter(response, "PubComment");
    
    if ((toStr == null) || (toStr.equals("")) ||
        (itemStr == null) || (itemStr.equals(""))||
        (name == null) || (name.equals(""))||
        (pass == null) || (pass.equals("")))
    {
      printError("User id, name and password are required - Cannot process the request<br>", sp);
      return ;
    }

    Context initialContext = null;
    try
    {
      initialContext = new InitialContext();
    } 
    catch (Exception e) 
    {
      printError("Cannot get initial context for JNDI: " + e+"<br>", sp);
      return ;
    }

    // Authenticate the user who want to comment
    Auth auth = new Auth(initialContext, sp);
    int userId = auth.authenticate(name, pass);
    if (userId == -1)
    {
      printError("You don't have an account on RUBiS!<br>You have to register first.<br>", sp);
      return ;	
    }

    // Try to find the user corresponding to the 'to' ID
    UserHome uHome;
    ItemHome iHome;
    try 
    {
      uHome = (UserHome)PortableRemoteObject.narrow(initialContext.lookup("UserHome"),
                                                    UserHome.class);
      iHome = (ItemHome)PortableRemoteObject.narrow(initialContext.lookup("ItemHome"),
                                                    ItemHome.class);
    } 
    catch (Exception e)
    {
      printError("Cannot lookup User or Item: " +e+"<br>", sp);
      return ;
    }

    try
    {
      Integer toId = new Integer(toStr);
      Integer itemId = new Integer(itemStr);
      User    to = uHome.findByPrimaryKey(new UserPK(toId));
      Item    item = iHome.findByPrimaryKey(new ItemPK(itemId));
      String  toName = to.getNickName();

      // Display the form for comment
      sp.printHTMLheader("RUBiS: Comment service");
      sp.printHTML("<center><h2>Give feedback about your experience with "+toName+"</h2><br>\n");
      sp.printHTML("<form action=\"/servlet/edu.rice.rubis.beans.servlets.StoreComment\" method=POST>\n"+
                   "<input type=hidden name=to value="+toStr+">\n"+
                   "<input type=hidden name=from value="+userId+">\n"+
                   "<input type=hidden name=itemId value="+itemId+">\n"+
                   "<center><table>\n"+
                   "<tr><td><b>From</b><td>"+name+"\n"+
                   "<tr><td><b>To</b><td>"+toName+"\n"+
                   "<tr><td><b>About item</b><td>"+item.getName()+"\n"+
                   "<tr><td><b>Rating</b>\n"+
                   "<td><SELECT name=rating>\n"+
                   "<OPTION value=\"5\">Excellent</OPTION>\n"+
                   "<OPTION value=\"3\">Average</OPTION>\n"+
                   "<OPTION selected value=\"0\">Neutral</OPTION>\n"+
                   "<OPTION value=\"-3\">Below average</OPTION>\n"+
                   "<OPTION value=\"-5\">Bad</OPTION>\n"+
                   "</SELECT></table><p><br>\n"+
                   "<TEXTAREA rows=\"20\" cols=\"80\" name=\"comment\">Write your comment here</TEXTAREA><br><p>\n"+
                   "<input type=submit value=\"Post this comment now!\"></center><p>\n");
    }
    catch (Exception e)
    {
      printError("This item does not exist (got exception: " +e+")<br>", sp);
      return ;
    }

    sp.printHTMLfooter();
  }

  /**
   * Call the <code>doGet</code> method.
   *
   * @param request a <code>HttpServletRequest</code> value
   * @param response a <code>HttpServletResponse</code> value
   * @exception IOException if an error occurs
   * @exception ServletException if an error occurs
   */
  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException
  {
    doGet(request, response);
  }
}

File: RUBiS-1.4.3/PHP/PutComment.php
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
  <body>
    <?php
    $scriptName = "PutComment.php";
    include("PHPprinter.php");
    $startTime = getMicroTime();
    
    $to = $HTTP_POST_VARS['to'];
    if ($to == null)
    {
      $to = $HTTP_GET_VARS['to'];
      if ($to == null)
      {
         printError($scriptName, $startTime, "PutComment", "You must provide a user identifier!<br>");
         exit();
      }
    }      

    $nickname = $HTTP_POST_VARS['nickname'];
    if ($nickname == null)
    {
      $nickname = $HTTP_GET_VARS['nickname'];
      if ($nickname == null)
      {
         printError($scriptName, $startTime, "PutComment", "You must provide a nick name!<br>");
         exit();
      }
    }

    $password = $HTTP_POST_VARS['password'];
    if ($password == null)
    {
      $password = $HTTP_GET_VARS['password'];
      if ($password == null)
      {
         printError($scriptName, $startTime, "PutComment", "You must provide a password!<br>");
         exit();
      }
    }

    $itemId = $HTTP_POST_VARS['itemId'];
    if ($itemId == null)
    {
      $itemId = $HTTP_GET_VARS['itemId'];
      if ($itemId == null)
      {
         printError($scriptName, $startTime, "PutComment", "You must provide an item identifier!<br>");
         exit();
      }
    }

    getDatabaseLink($link);

    beginRO($link);
    // Authenticate the user
    $userId = authenticate($nickname, $password, $link);
    if ($userId == -1)
    {
      printError($scriptName, $startTime, "PutComment", "You don't have an account on RUBiS!<br>You have to register first.<br>\n");
      exit();	
    }

function putcomment($link, $to, $itemId) {
    $row = getItem($link, $itemId);
    if (!$row)
    {
      global $scriptName, $startTime;
      printError($scriptName, $startTime, "PutComment", "<h3>Sorry, but this item does not exist.</h3><br>");
      return NULL;
    }

    $userRow = getUser($link, $to);
    if (!$userRow)
    {
      global $scriptName, $startTime;
      printError($scriptName, $startTime, "PutComment", "<h3>Sorry, but this user does not exist.</h3><br>");
      return NULL;
    }

    printHTMLheader("RUBiS: Comment service");

    print("<center><h2>Give feedback about your experience with ".$userRow["name"]."</h2><br>\n");
    $out = ("<form action=\"/PHP/StoreComment.php\" method=POST>\n".
          "<input type=hidden name=to value=$to>\n".
          "<input type=hidden name=from value=\$XXXuserIdXXX>\n".
          "<input type=hidden name=itemId value=$itemId>\n".
          "<center><table>\n".
          "<tr><td><b>From</b><td>\$XXXnicknameXXX\n".
          "<tr><td><b>To</b><td>".$userRow["nickname"]."\n".
          "<tr><td><b>About item</b><td>".$row["name"]."\n".
          "<tr><td><b>Rating</b>\n".
          "<td><SELECT name=rating>\n".
          "<OPTION value=\"5\">Excellent</OPTION>\n".
          "<OPTION value=\"3\">Average</OPTION>\n".
          "<OPTION selected value=\"0\">Neutral</OPTION>\n".
          "<OPTION value=\"-3\">Below average</OPTION>\n".
          "<OPTION value=\"-5\">Bad</OPTION>\n".
          "</SELECT></table><p><br>\n".
          "<TEXTAREA rows=\"20\" cols=\"80\" name=\"comment\">Write your comment here</TEXTAREA><br><p>\n".
          "<input type=submit value=\"Post this comment now!\"></center><p>\n");

    return $out;
}

if (MICROCACHE)
  $page = putcomment($link, $to, $itemId);
else
    $page = wrap(true, 'putcomment', $link, $to, $itemId);
if (!$page)
  exit();
echo str_replace(array('$XXXuserIdXXX', '$XXXnicknameXXX'), array($userId, $nickname), $page);

    commit($link);
    sql_close($link);
    
    printHTMLfooter($scriptName, $startTime);
    ?>
  </body>
</html>
